Add the ability to infer assets from strings for Python

thejcannon opened the pull request:
# High level

Although less structured, it shouldn't be difficult to infer assets from strings and filter the noise to only what is valid.

The use-case should be evident, however here's a few intentional design decisions:
- Ability to infer both resources (assumed to be relative to the module for now) and files (relative to the build root)
- Scope the possible strings to a subset of what really is possible. On Unix, basically anything is a valid filename, but that could be needless explosive.
- Use strings (as opposed to just parsing `pkgutil.get_data`) because some people use custom resource types. E.g. `Resource("data/foo.txt")`. We can add `pkgutil.get_data` support later.

Example:

Given:
```
<root>/
├─ configs/
│  ├─ prod.txt (a file)
├─ src/
│  ├─ app1/
│  │  ├─ data/
│  │  │  ├─ db.json (a resource)
│  │  ├─ app.py
```
(With all the right `BUILD` file plumbing)
Then `src/app1/app.py` containing code like `open("configs/prod.txt") as envvars` or `Database("data/db.json")` would have the relevant file/resource dependency inferred by Pants! :tada:

thejcannon commented on src/python/pants/backend/python/dependency_inference/dependency_parser.py:
```diff
@@ -14,7 +19,7 @@
 
-# NOTE: Must call .format(min_dots=X) on this string to use it.
+# NOTE: Must call .format(import_min_dots=X, resource_min_slashes=Y) on this string to use it.
 _SCRIPT_FORMAT = """\
```
At this point I think it's worth investigating (in a different PR) getting this in it's own file. It needs syntax highlighting and to be linted/formatted/etc...

thejcannon replied:
https://github.com/pantsbuild/pants/pull/14080

thejcannon commented on src/python/pants/backend/python/dependency_inference/rules.py:
```diff
@@ -40,6 +44,33 @@
 
 logger = logging.getLogger(__name__)
 
+# @TODO: Find a home for these
```
Opening issue for TODO

thejcannon replied:
This isn't unique to Python at all, but there isn't a utility bucket. Maybe there should be?

thejcannon replied:
This isn't unique to Python at all, but there isn't a utility bucket. Maybe there should be?

chrisjrn replied:
I've proposed `backend/common` as a place for things like this; discussing over on slack now :)

thejcannon replied:
For posterity: https://pantsbuild.slack.com/archives/C0D7TNJHL/p1641577800027600

thejcannon replied:
`core/util_rules` seems to be the answer :tada:

thejcannon commented on src/python/pants/backend/python/dependency_inference/rules.py:
```diff
@@ -167,34 +218,53 @@ async def infer_python_dependencies_via_imports(
         ),
     )
 
+    inferred_deps: set[Address] = set()
+
+    detected_imports = detected_dependencies.imports
+    detected_resources = detected_dependencies.resources
+    resources_by_path: Dict[pathlib.Path, Target] = {}
+    for file_tgt in all_resource_targets.files:
+        resources_by_path[pathlib.Path(file_tgt[FileSourceField].file_path)] = file_tgt
+    for resource_tgt in all_resource_targets.resources:
+        path = pathlib.Path(resource_tgt[ResourceSourceField].file_path)
+        resources_by_path[path] = resource_tgt
+
+    for pkgname, filepath in detected_resources:
+        resource_path = pathlib.Path(*pkgname.split(".")).parent / filepath
+        inferred_resource_tgt = resources_by_path.get(resource_path)
+        if inferred_resource_tgt:
+            inferred_deps.add(inferred_resource_tgt.address)
```
Might split this into a helper...

stuhood commented:
> Can someone re-kick the tests?

Will need to rebase I think: the fix for this landed on `main` as 5c1016945c1c0ae75e424601edf83f37353afb3a.

thejcannon commented:
I suppose rebasing is an easy re-kick. :raised_hands:

thejcannon commented on src/python/pants/backend/python/dependency_inference/rules.py:
```diff
@@ -89,6 +121,18 @@ def register_options(cls, register):
+            "--string-resources",
+            default=False,
+            type=bool,
+            help=("@TODO"),
+        )
+        register(
+            "--string-resources-min-slashes",
+            default=1,
+            type=int,
+            help=("@TODO"),
```
Opening issue for TODOs

thejcannon commented on src/python/pants/backend/python/dependency_inference/BUILD:
```diff
@@ -1,13 +1,6 @@
 # Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 
-python_sources(
-    overrides={
-        "parse_python_imports.py": {
-            # This Python script is loaded as a resource, see parse_python_imports.py for more info.
-            "dependencies": ["./scripts:import_parser"]
-        }
-    }
-)
```
You might ask where this has gone. The answer: it's being inferred :magic_wand:

thejcannon commented on src/python/pants/backend/python/dependency_inference/rules.py:
```diff
@@ -225,7 +295,7 @@ async def infer_python_dependencies_via_imports(
                 "One or more unowned dependencies detected. Check logs for more details."
```
(I can't seem to open an issue on line 282 :angry: )

If I'm being a pedant, the name `unowned_dependency_behavior` is now misleading, as files/resources are dependencies, but we just toss the inferred ones that don't have an owner.

A more correct name would be `unowned_import_behavior`. But also :man_shrugging:

kaos commented on src/python/pants/backend/python/dependency_inference/rules.py:
```diff
@@ -40,6 +44,33 @@
+        if isinstance(tgt, ResourceTarget):
+            resources.append(tgt)
+        if isinstance(tgt, FileTarget):
+            files.append(tgt)
```
I think we usually prefer to look at the fields, rather than the target type.
> Often, you may want to see if a target type has a particular Field registered. This is useful to filter targets. Use the methods .has_field() and .has_fields().
From https://www.pantsbuild.org/v2.9/docs/rules-api-and-target-api

So in this case, look for targets that has the `ResourceSourceField` and `FileSourceField` respectively. That way, you won't exclude valid other potentially valid targets..

kaos reviewed (commented):
Looking good. Haven't worked through it all in details, though.. :)

kaos commented on src/python/pants/backend/python/dependency_inference/scripts/dependency_parser.py:
```diff
@@ -114,7 +128,14 @@ def main(filename):
+                "resources": list(sorted(visitor.resources)),
+            }
+        ).encode("utf8")
+    )
```
This is a much better idea than what I've done in the `dockerfile_parser`.. will def steal this one :)

thejcannon commented on src/python/pants/backend/python/dependency_inference/scripts/dependency_parser.py:
```diff
@@ -18,13 +18,22 @@
 )
+# This regex is used to infer resource names from strings, e.g.
+#  `load_resource("data/db1.json")
+# Since Unix allows basically anything for filenames, we require some "sane" subset of possibilities
+#  namely, word-character filenames and a mandatory extension.
+STRING_RESOURCE_REGEX = re.compile(
+    r"^([\w]*\/){" + MIN_SLASHES + r",}\w*(\.[^\/\.]+)+$",
+    re.UNICODE,
+)
```
Open discussion: What's the harm in just matching every string?
- Potentially a little bit more of work to see if a string is a valid resource
- Over-matching. Is that really an issue for resources though? :thinking:

benjyw replied:
There are a lot of string literals in code, so it might be a lot more work?

thejcannon replied:
We do the work in one place or another. In this case it is in matching regexes. But could be shifted into Pants.

Without numbers though, I'd believe the overhead is negligible.

The benefit is not scoping us out of extensionless files, or files in the same dir as the module.

thejcannon replied:
Another dimension here is I think users should be able to tweak `python-infer` at the `BUILD` level. So instead of dragging the overhead of string-matching-resources for the entire monorepo, it'd just be enabled for the specific files.

thejcannon replied:
I suppose we can also futher configure this with a `allow_extenionless` flag...

Eric-Arellano commented on src/python/pants/backend/python/dependency_inference/rules.py:
```diff
@@ -162,38 +189,65 @@ class InferPythonImportDependencies(InferDependenciesRequest):
     request: InferPythonImportDependencies,
     python_infer_subsystem: PythonInferSubsystem,
     python_setup: PythonSetup,
+    all_resource_targets: AllResourcesTargets,
```
This is slow to compute, and it's wasted computation if you don't have resource inference enabled. Until we add https://github.com/pantsbuild/pants/issues/12946, the solution is to add `AllResourceAndFileTargetsRequest` and then use `Get(AllResourceAndFileTargets, AllResourceAndFileTargetsRequest())` :/

thejcannon replied:
Actually this had made me realize, we might want a union rule for this, and split out the implementation into it's own rule. That also means maybe splitting the resource parsing out of the helper file into its own file.
FWIW this method is called `infer_python_dependencies_via_imports`, which is incorrect for this use case.

Thoughts?

Eric-Arellano reviewed (commented):
Neat! Iirc, this would not handle this: `importlib.resources.read_text("path.to.module", "foo.txt")`, right? Even if not, this is still great.

thejcannon commented:
> Neat! Iirc, this would not handle this: `importlib.resources.read_text("path.to.module", "foo.txt")`, right? Even if not, this is still great.

It does not. I didn't want to risk adding too many things to this PR, and import inference doesn't support `importlib.import_module()`. So I figured we can add support for both of these later.

Eric-Arellano commented:
> So I figured we can add support for both of these later.

Makes sense. Only thing we need to be careful about is how to handle backwards-compatibility. If we change the behavior of this new option to become more powerful / infer more than before, that is somewhat a breaking change. I think it's acceptable, but we should probably hedge in the `help` message by saying

> Note that we hope to continue improving the inference of resources/files, such as eventually inferring `importlib.resources` statements. These improvements will happen without the normal deprecation policy.

thejcannon commented:
> > So I figured we can add support for both of these later.
> 
> Makes sense. Only thing we need to be careful about is how to handle backwards-compatibility. If we change the behavior of this new option to become more powerful / infer more than before, that is somewhat a breaking change. I think it's acceptable, but we should probably hedge in the `help` message by saying
> 
> > Note that we hope to continue improving the inference of resources/files, such as eventually inferring `importlib.resources` statements. These improvements will happen without the normal deprecation policy.

That's likely to be true of `imports` as well, right? With `importlib.import_module`?

thejcannon commented:
> Actually this had made me realize, we might want a union rule for this, and split out the implementation into it's own rule. That also means maybe splitting the resource parsing out of the helper file into its own file.
FWIW this method is called infer_python_dependencies_via_imports, which is incorrect for this use case.

Yeah OK I think this is likely the route to take. I'll most likely post a new PR for that since the changes will be somewhat drastic. :crossed_fingers: hopefully there isn't too much duplication between this and import parsing.

Eric-Arellano commented:
I don't think I agree with splitting out the parsing into two separate `Process`es — there is overhead to launching more Processes. Notably, they will both have the exact same `input_digest`, so there's no caching benefit to this.

Instead, it's important that the parser do the minimum work necessary. If you don't care about resources, then don't try to parse them. It looks like you already made that optimization.

> we might want a union rule for this

That sounds like it might be more complex. What would be the benefit?

thejcannon commented:
It's more about simplification and separation of concerns. The `infer_python_dependencies_via_imports` (which needs to be renamed, oops) already does a lot for python imports, and now we're adding more into it.

Perhaps splitting it into helpers (not necessarily needing #14238 but that'd be even nicer) would qualm some doubts.

For implementations sake, you might be right that bundling them together is still simpler.

Eric-Arellano commented:
> It's more about simplification and separation of concerns.

I didn't think this PR was too complex - I could follow it pretty well. Re separation of concerns, these concerns _are_ highly related imo.

thejcannon commented on src/python/pants/backend/python/dependency_inference/rules.py:
```diff
@@ -89,6 +96,18 @@ def register_options(cls, register):
             ),
         )
+        register(
+            "--string-resources-and-files",
```
@Eric-Arellano what do we think about this being `resources-and-files-from-strings`. Otherwise it sounds like there'd be non-string resource and files parsed :thinking:

Eric-Arellano replied:
Yeah I like that a lot

thejcannon replied:
I'm still not sitting on this well. Opened a discussion in Slack: https://pantsbuild.slack.com/archives/C0D7TNJHL/p1643421103775879

thejcannon commented:
Reabsed on `main` and no longer a WIP

thejcannon commented on src/python/pants/backend/python/dependency_inference/rules.py:
```diff
@@ -242,7 +278,73 @@ async def infer_python_dependencies_via_imports(
+    unowned_imports: set[str] = set()
+    parsed_imports = parsed_dependencies.imports
+    parsed_resources = parsed_dependencies.resources
+    if not python_infer_subsystem.imports:
+        parsed_imports = ParsedPythonImports([])
+
+    if parsed_imports:
```
This is awkward. As far as settings goes we have `imports`, `string-imports` (which is ignored is `imports` is `False`) and this new one.

I'm wondering if we should pass `imports` into the `dependency_parser`, so it becomes the one-and-only funnel.

thejcannon commented on src/python/pants/backend/python/dependency_inference/rules.py:
```diff
@@ -96,6 +109,26 @@ def register_options(cls, register):
                 "treated as a potential dependency if this option is set to 2 but not if set to 3."
             ),
         )
+        register(
+            "--assets",
```
This makes me want "sub" subsystems.
E.g. `python-infer.assets` `python-infer.imports`, etc...

thejcannon commented on src/python/pants/backend/python/dependency_inference/parse_python_deps.py:
```diff
@@ -30,22 +33,48 @@ class ParsedPythonImportInfo:
+
+
+def _filepath_to_modname(filepath: str) -> str:
+    return str(pathlib.Path(filepath).with_suffix("")).replace("/", ".")
```
I'm going to (lightly) punt on this, and tackle it in a follow-up PR. This pattern exists in a few files.

thejcannon commented:
OK I did another once-over on the PR. Should be good to re-review. Let me know if it needs to be split.

Eric-Arellano commented on src/python/pants/backend/python/dependency_inference/scripts/dependency_parser.py:
```diff
@@ -79,6 +87,12 @@ def find_token(string):
                 find_token(alias.asname)
 
+    def maybe_add_string_dependency(self, node, s):
+        if STRING_IMPORTS and STRING_IMPORT_REGEX.match(s):
+            self.weak_imports.setdefault(s, node.lineno)
+        if ASSETS and ASSET_REGEX.match(s):
+            self.assets.add((None, s))
+
```
The only time this is possible is if you set min slashes and min dots to 0, right? Otherwise, this could early return? Given that this regex runs against _every single string_ and regex is relatively fast but does have a perf hit, it might be worth adding that eager exit. Wdyt?

thejcannon replied:
(Note the file we're parsing has a bunch of strings)
on `main`:
```
Benchmark 1: STRING_IMPORTS=Y MIN_DOTS=2 python src/python/pants/backend/python/dependency_inference/scripts/import_parser.py src/python/pants/option/global_options.py
  Time (mean ± σ):      33.3 ms ±   1.9 ms    [User: 27.7 ms, System: 5.6 ms]
  Range (min … max):    28.4 ms …  41.1 ms    100 runs
```

with my changes:
```
Benchmark 1: STRING_IMPORTS=Y STRING_IMPORT_MIN_DOTS=2 ASSETS=y ASSET_MIN_SLASHES=1 python src/python/pants/backend/python/dependency_inference/scripts/dependency_parser.py src/python/pants/option/global_options.py
  Time (mean ± σ):      33.4 ms ±   1.7 ms    [User: 28.5 ms, System: 4.9 ms]
  Range (min … max):    28.5 ms …  36.8 ms    100 runs
```

Eric-Arellano replied:
Ah okay cool. Nevermind me :) thanks

Eric-Arellano commented on src/python/pants/backend/python/dependency_inference/rules.py:
```diff
@@ -182,10 +207,19 @@ async def infer_python_dependencies_via_imports(
+
+def _maybe_warn_unowned(
+    address: Address,
+    file: str,
```
That's a builtin. Maybe `file_path`?

thejcannon replied:
Not anymore... https://docs.python.org/3/library/functions.html#built-in-funcs

```
joshuacannon@CEPHANDRIUS:~$ python
Python 3.8.12 (default, Sep 10 2021, 00:16:05) 
[GCC 7.5.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> file
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'file' is not defined
```

Eric-Arellano commented on src/python/pants/backend/python/dependency_inference/rules.py:
```diff
@@ -203,7 +237,66 @@ async def infer_python_dependencies_via_imports(
+        inferred_deps.update(import_deps)
+
+    if parsed_assets:
+        all_asset_targets = await Get(AllAssetTargets, AllAssetTargetsRequest())
```
Yay thanks for making this lazy.

Eric-Arellano reviewed (dismissed):
🥳

Eric-Arellano commented on src/python/pants/backend/python/dependency_inference/rules.py:
```diff
@@ -130,41 +161,35 @@ class InferPythonImportDependencies(InferDependenciesRequest):
+    all_asset_targets: AllAssetTargets,
+    assets: ParsedPythonAssets,
+) -> Iterator[Address]:
+    assets_by_path: Dict[PurePath, Target] = {}
+    for file_tgt in all_asset_targets.files:
+        assets_by_path[PurePath(file_tgt[FileSourceField].file_path)] = file_tgt
+    for resource_tgt in all_asset_targets.resources:
+        path = PurePath(resource_tgt[ResourceSourceField].file_path)
+        assets_by_path[path] = resource_tgt
```
This can be slow to create, so I think we want it in a dedicated rule so that we get memoization. Basically `module_mapper.py` for resources. Also...this should probably handle ambiguity similar to Python's `module_mapper.py` - you can't infer if >1 target exposes the same "path"

thejcannon replied:
What makes you think it'd be slow? It's pure-python dictionary population so I think it'd be in the noise here.

thejcannon replied:
My rough timing in the Pants repo shows it to be less than a second

Eric-Arellano replied:
Iterating over `AllTargets` in a repo with 10,000 targets. Users have reported that the Python module mapping creation can take several seconds. Better to memoize that.

Beyond the perf, I do think the ambiguity part is important. We might not want to do the normal warning with ambiguity because of too many false positives, but we at least do want to not infer a dep when ambiguous.

Eric-Arellano replied:
Note that you're re-creating the module mapping for every single target: the rule that calls `_get_inferred_asset_deps` is executed every single Python target. This should be memoized through a dedicated rule to create the module mapping.

thejcannon replied:
> Note that you're re-creating the module mapping for every single target

:facepalm: Duh...

Eric-Arellano reviewed (commented):
> Ability to infer both resources (assumed to be relative to the module, unless pkgutil.get_data is used) and files

Is the PR description still up-to-date? I'm not following the code for this part.

Generally, it seems that this only infers in the format `path/to/f.txt`, _not_ `path.to.my_resource`, right?

thejcannon commented:
> > Ability to infer both resources (assumed to be relative to the module, unless pkgutil.get_data is used) and files
> 
> Is the PR description still up-to-date? I'm not following the code for this part.
> 
> Generally, it seems that this only infers in the format `path/to/f.txt`, _not_ `path.to.my_resource`, right?

Fixed the wording

Eric-Arellano commented on src/python/pants/backend/python/dependency_inference/parse_python_dependencies.py:
```diff
@@ -31,21 +34,44 @@ class ParsedPythonImports(FrozenDict[str, ParsedPythonImportInfo]):
+class ParsedPythonAssets(DeduplicatedCollection[Tuple[str, str]]):
+    """All the discovered possible assets from a Python source file.
+
+    The tuple is of (containing module, filename).
```
Ah I didn't grok this well enough and it resulted in some of my confusion. I think an example could help, e.g. `("path.to.module", "foo.txt")`.

Eric-Arellano commented on src/python/pants/backend/python/dependency_inference/parse_python_dependencies.py:
```diff
@@ -61,23 +87,38 @@ async def parse_python_imports(request: ParsePythonImportsRequest) -> ParsedPyth
+        assets=ParsedPythonAssets(
+            [
+                (
+                    _filepath_to_modname(request.source.file_path) if pkgname is None else pkgname,
```
`_filepath_to_modname` gets called repeatedly with the same value. Probably better to save as a variable.

Eric-Arellano commented on src/python/pants/backend/python/dependency_inference/parse_python_dependencies.py:
```diff
@@ -31,21 +34,44 @@ class ParsedPythonImports(FrozenDict[str, ParsedPythonImportInfo]):
+
+
+def _filepath_to_modname(filepath: str) -> str:
+    return str(pathlib.Path(filepath).with_suffix("")).replace("/", ".")
```
We try to limit ourselves to `PurePath` to avoid the risk of using `Path`'s IO side effects.

Eric-Arellano commented on src/python/pants/backend/python/dependency_inference/scripts/dependency_parser.py:
```diff
@@ -79,6 +87,12 @@ def find_token(string):
+        if STRING_IMPORTS and STRING_IMPORT_REGEX.match(s):
+            self.weak_imports.setdefault(s, node.lineno)
+        if ASSETS and ASSET_REGEX.match(s):
+            self.assets.add((None, s))
```
This is always set to `None` for now, but I think you intend to eventually infer the `pkg_name`? I recommend deleting that part of the code for now. This script is an implementation detail and we can add the pkg_name code whenever. I think this will simplify callers a bit.

Eric-Arellano commented on src/python/pants/backend/python/dependency_inference/parse_python_dependencies.py:
```diff
@@ -31,21 +32,37 @@ class ParsedPythonImports(FrozenDict[str, ParsedPythonImportInfo]):
     """All the discovered imports from a Python source file mapped to the relevant info."""
 
 
+class ParsedPythonAssets(DeduplicatedCollection[str]):
```
Maybe `ParsedPythonAssetPaths`? I continue to confuse myself with what this is, I think because of the old setup of it being a tuple.

thejcannon replied:
Yeah. If/when we support `pkgutil.get_data` we'll be changing this name :sweat_smile: but for now that works!

Eric-Arellano commented on src/python/pants/backend/python/dependency_inference/rules.py:
```diff
@@ -89,6 +103,24 @@ class PythonInferSubsystem(Subsystem):
+        ),
+    )
+    assets_min_slashes = IntOption(
+        "--assets-min-slashes",
```
I'm thinking about imports in the form `pkgutil.get_data("path.to.module", "foo.txt")` and trying to figure out how we will want to model the options. It's fine even if we never implement it, but I suspect we may want to.

We'd probably want `--assets-min-dots`? I think it's plausible you'd want to have a different value for than that slashes? 

I was wondering if we could throw this into `--string-imports-min-dots`, but I don't think these are string imports. We need to preserve the file extension of `foo.txt`, whereas string imports are _Python_ imports.

thejcannon replied:
So this is getting into implementation details of a future change, so I wouldn't worry too much here.

The way I see it is that `pkgutil.get_data("path.to.module", "foo.txt")` is unambiguous, so wouldn't be subject to `min-dots` or `min-slashes` (but probably still subject to `assets`).

Eric-Arellano replied:
> So this is getting into implementation details of a future change, so I wouldn't worry too much here.

I'm worried about it only so far as figuring out what option we likely want to have. It'd be great to avoid having to deprecate an option in the future.

--

The idea with min-slashes & min-dots is to help counteract false positives btw. We saw with pantsbuild.pants that setting too low caused way too many things to get inferred. 

--

I think we probably do want `--assets-min-dots`..If that's the case, then we're good to go. No need to think more about that on this PR.

thejcannon replied:
FWIW I think with some of the improvements (strong v weak imports and resolves) that technically those options aren't needed as much since the false positives are minimized even when running on all strings. But not everyone is using the right mix of options. (Thats for imports, at least)

In this case (and imports via function) I see the functional dependency (using the Pythonic function) as unambiguous. If you're not loading an asset via `pkgutil.get_data` what else are you doing?

Also, I don't really see where `--asset-min-dots` fits in :/ The only time we'd infer an asset and need dots is in `pkgutil.get_data` which is unambigous.

Eric-Arellano replied:
From some DMs, there's some confusion here still about what I'm trying to say. 

But it's okay, I've thought it more and I think we would want two options. Meaning, no changes necessary to this option in this PR. We're good.

Eric-Arellano commented on src/python/pants/backend/python/dependency_inference/rules.py:
```diff
@@ -89,6 +103,24 @@ class PythonInferSubsystem(Subsystem):
     )
+    assets = BoolOption(
+        "--assets",
+        default=False,
```
If you didn't have to consider our deprecation policy, should this default to True? We don't have string imports default to True because of false positives. Not sure what the right behavior is here.

thejcannon replied:
Yeah. OK. I can get very behind making `True` the default.

Eric-Arellano replied:
Ah, but note that we do need to consider our deprecation policy, I think. Possibly. Unclear to me. I was only asking what the aspiration is.

In the interest of landing this for 2.10, how about using `False` for now, having a conversation in Slack or dedicated PR on Monday, and then possibly changing it in time for 2.10? That'd be good anyways to have two entries in changelog:

```
# new feature
* Add `[python-infer].assets` to infer resources and files from strings

# user-api changes
* Enable `[python-infer].assets` by default
```

thejcannon commented:
Updated with:
- Asset inference on by default
- Reverse mapping of assets by path in a rule
- Added file inference. That was the intent from the start, but for whatever reason I didn't test it (and the code wasn't good)

Eric-Arellano commented on src/python/pants/backend/python/dependency_inference/rules.py:
```diff
@@ -130,41 +155,43 @@ class InferPythonImportDependencies(InferDependenciesRequest):
+    explicitly_provided_deps: ExplicitlyProvidedDependencies,
+) -> Iterator[Address]:
+    for filepath in assets:
+        resource_path = PurePath(request_file_path).parent / filepath
+        inferred_resource_tgts = assets_by_path.resources.get(resource_path, frozenset())
+        inferred_file_tgts = assets_by_path.files.get(PurePath(filepath), frozenset())
```
I think it'd be helpful to have a comment explaining why resources only use the `resource_path` and not the full path. But why `file` uses the full path. I originally was envisioning that we look up both mappings with both path types.

I think it makes sense to me why you're doing it this more limited way, but isn't super obvious to me either.

thejcannon replied:
Oh hmm that's interesting... Yeah I'll stew on this a bit and make this very clear in a comment. Very good point.

Eric-Arellano replied:
Note that dictionary lookups are fast, so it doesn't seem wrong to me to do both lookups?

thejcannon replied:
Ok yeah, you made me think about it harder and I'm confident this approach is the right one.

Let me know if the comment needs any tuning.

Eric-Arellano reviewed (commented):
Thank you for iterating on this! Looks good once the `default` is reverted to not break our deprecation policy.

Eric-Arellano reviewed (approved):
🎉
